src/recap.c

```diff
@@ -0,0 +1,490 @@
+  fprintf(stdout,
+          "\t-d\t\t Verbosity level (use multiple -dd... to increase)\n");
+  fprintf(stdout, "\t-h\t\t Show help\n");
+  fprintf(stdout, "\t-v\t\t Show app version\n\n");
+  fprintf(stdout, "Copyright NQMCyber Ltd\n\n");
+  exit(EXIT_SUCCESS);
+}
+
+/* Diagnose an error in command-line arguments and
+   terminate the process */
+void log_cmdline_error(const char *format, ...) {
+  va_list argList;
+
+  fflush(stdout); /* Flush any pending stdout */
+
+  fprintf(stdout, "Command-line usage error: ");
+  va_start(argList, format);
+  vfprintf(stdout, format, argList);
+  va_end(argList);
+
+  fflush(stderr); /* In case stderr is not line-buffered */
+  exit(EXIT_FAILURE);
+}
+
+void process_app_options(int argc, char *argv[], uint8_t *verbosity,
+                         char **pcap_path, char **db_path, char **ifname) {
+  int opt;
+
+  while ((opt = getopt(argc, argv, OPT_STRING)) != -1) {
+    switch (opt) {
+      case 'h':
+        show_app_help(argv[0]);
+        break;
+      case 'v':
+        show_app_version();
+        exit(EXIT_SUCCESS);
+        break;
+      case 'p':
+        *pcap_path = os_strdup(optarg);
+        break;
+      case 'f':
+        *db_path = os_strdup(optarg);
+        break;
+      case 'i':
+        *ifname = os_strdup(optarg);
+        break;
+      case 'd':
+        (*verbosity)++;
+        break;
+      case ':':
+        log_cmdline_error("Missing argument for -%c\n", optopt);
+        break;
+      case '?':
+        log_cmdline_error("Unrecognized option -%c\n", optopt);
+        break;
+      default:
+        show_app_help(argv[0]);
+    }
+  }
+}
+
+ssize_t read_pcap_stream_fd(struct pcap_stream_context *pctx, size_t len,
+                            char **data) {
+  if ((*data = os_malloc(len)) == NULL) {
+    log_errno("os_malloc");
+    return -1;
+  }
+
+  return (ssize_t)fread(*data, sizeof(char), len, pctx->pcap_fd);
+}
+
+ssize_t read_pcap(struct pcap_stream_context *pctx, size_t len) {
+  char *data = NULL;
+  ssize_t read_size, current_size;
+
+  if ((read_size = read_pcap_stream_fd(pctx, len, &data)) < 0) {
+    log_error("read_pcap_stream_fd fail");
+    return -1;
+  }
+
+  pctx->total_size += read_size;
+
+  current_size = read_size + pctx->data_size;
+
+  if (read_size > 0) {
+    if ((pctx->pcap_data = os_realloc(pctx->pcap_data, current_size)) == NULL) {
+      log_errno("os_realloc");
+      os_free(data);
+      return -1;
+    }
+    os_memcpy(&pctx->pcap_data[pctx->data_size], data, read_size);
+    pctx->data_size = current_size;
+  }
+  os_free(data);
+
+  return read_size;
+}
+
+int process_pcap_header_state(struct pcap_stream_context *pctx) {
+  ssize_t read_size = 0;
+  ssize_t pcap_header_size = (ssize_t)sizeof(struct pcap_file_header);
+
+  size_t len = (pcap_header_size > pctx->data_size)
+                   ? pcap_header_size - pctx->data_size
+                   : 0;
+
+  if ((read_size = read_pcap(pctx, len)) < 0) {
```

Comment:
Moving variable declaration right next to where the variable is used.

```suggestion
  ssize_t pcap_header_size = (ssize_t)sizeof(struct pcap_file_header);

  size_t len = (pcap_header_size > pctx->data_size)
                   ? pcap_header_size - pctx->data_size
                   : 0;

  ssize_t read_size = read_pcap(pctx, len);
  if (read_size < 0) {
```